### ISSUE TITLE
Why is LAMMPS_EXCEPTIONS option OFF by default?

### ISSUE THREAD

**Issue by yurivict** [lammps/lammps#2088]
https://github.com/lammps/lammps/blob/master/cmake/CMakeLists.txt#L195

Wouldn't it be better to enable them by default so that it would be possible to use lammps shared libs?

Thank you,
Yuri

**Comment by akohlmey** [comment]
Most people that use LAMMPS don't need (or want) that kind of behavior. You can use LAMMPS as a (shared) library just fine without. And, of course, you are free to enable it and compile LAMMPS that way you want.

**Comment by yurivict** [comment]
> Most people that use LAMMPS don't need (or want) that kind of behavior.

What are the downsides of exceptions that they don't like?

> You can use LAMMPS as a (shared) library just fine without.

But it would terminate the user's process with ```exit(1)``` when such condition occurs. This isn't necessarily the desired behavior.

------

It seems that default (no exception) branches in the code are different every time, so it's unclear how can the user know what action to perform in case exceptions occur.

**Comment by akohlmey** [comment]
> > Most people that use LAMMPS don't need (or want) that kind of behavior.
> 
> What are the downsides of exceptions that they don't like?

Most error conditions you cannot cleanly recover from. Keep in mind that LAMMPS is first and foremost an MPI parallel code and with exception handling enabled you get back to the calling code, but not necessarily on all MPI ranks. Thus to recover from a fatal condition you almost always have to restart the application from scratch as you need to reset the communication layer.

Also most people use LAMMPS as a standalone binary and that is where there is no difference in behavior, only the code path is different.

> > You can use LAMMPS as a (shared) library just fine without.
> 
> But it would terminate the user's process with `exit(1)` when such condition occurs. This isn't necessarily the desired behavior.

see above. and - let me remind you again - you are free to enable it. that is why it is an option. you have the choice. if you want it, take it.
 
> It seems that default (no exception) branches in the code are different every time, so it's unclear how can the user know what action to perform in case exceptions occur.

the default action to terminate the application is almost always what you want.
we added the exception handling for the special case of a graphical application where then a non-MPI version of LAMMPS was linked and not just run as a child process. in that case, a failure in the input, for example, would have the GUI crash. see https://github.com/ovilab/atomify and https://play.google.com/store/apps/details?id=com.compphys.atomify&hl=en_US or https://editor.lammps.org/

there are some other cases where people use a serial LAMMPS library from python with exceptions to do some unusual science, but for the most part when LAMMPS triggers an error, everything is messed up and you need to start over anyway.

**Comment by martok** [comment]
I kinda support this request, especially for shared object targets.
Alternatively (I haven't checked out the #1970 documentation yet), just document somewhere in "Build LAMMPS as a shared library" that it's a *really* good idea to use -DLAMMPS_EXCEPTIONS? It took me quite a while to discover this even exists originally.

> there are some other cases where people use a serial LAMMPS library from python with exceptions to do some unusual science, but for the most part when LAMMPS triggers an error, everything is messed up and you need to start over anyway.

It doesn't really need any unusual science. "Starting over" in a library context can just mean lammps_close'ing the faulted instance and lammps_open'ing a new one. Killing the host process is... let's call it highly irregular for a library.

**Comment by akohlmey** [comment]
> I kinda support this request, especially for shared object targets.
> Alternatively (I haven't checked out the #1970 documentation yet), just document somewhere in "Build LAMMPS as a shared library" that it's a _really_ good idea to use -DLAMMPS_EXCEPTIONS? It took me quite a while to discover this even exists originally.
> 
> > there are some other cases where people use a serial LAMMPS library from python with exceptions to do some unusual science, but for the most part when LAMMPS triggers an error, everything is messed up and you need to start over anyway.
> 
> It doesn't really need any unusual science. "Starting over" in a library context can just mean lammps_close'ing the faulted instance and lammps_open'ing a new one. Killing the host process is... let's call it highly irregular for a library.

that were cases i know about. 

I am OK with putting a note into the documentation. But I would really only call it a good idea to have exceptions enabled when not running in parallel. I don't want to change the default behavior, as that is what it has been historically for more than a decade. I also don't want to have a different behavior depending on whether shared or static or with MPI or without MPI or a combination. it just catches people by surprise. we just tried to simplify the build choices (and the number of permutations and commands/keywords to explain) by always building the library and linking the standalone executable from it, I don't want to add having to explain the different behavior and making things more complex again.

while it is growing (thanks to the improvements in the code and especially the python wrapper), the use of the LAMMPS library is still a minor segment of the LAMMPS use. so any global default setting should consider the majority of use cases. for me the benchmark to reconsider would be if MPI (or some other communication library replacing it) would have a way to automatically propagate exceptions to all MPI ranks and cancel/reset the state of all communicators.
outside of that, the only changes i would want to make (or accept as a contribution) are changes to the documentation.

**Comment by stanmoore1** [comment]
I agree with @akohlmey. For normal LAMMPS use (outside of library), if one hits an error there is no way to recover. Hard errors are hard errors for a reason, otherwise they should be warnings instead. How do you recover if the simulation box blew up, everything in simulation is full of NaNs, or the user gave invalid input?

**Comment by yurivict** [comment]
If errors are unrecoverable then you should remove ```LAMMPS_EXCEPTIONS``` completely because enabling it would lead to incorrect operation.

**Comment by akohlmey** [comment]
> If errors are unrecoverable then you should remove `LAMMPS_EXCEPTIONS` completely because enabling it would lead to incorrect operation.

This is a pointless argument because a research code like LAMMPS is **full** of pieces of code that, when using certain input, can lead to "incorrect operation". That is just the nature of a simulation software, especially one with so many features and so easy to extend and modify. There are too many permutations of combinations of features, that it is impossible to protect people from doing "bad" things. If people cannot handle that, that is not our problem. This is not a critical system software that people may be required to use to do important, unavoidable tasks.

In summary, I will see to review the documentation of compiling/using LAMMPS as a library and add  remarks about enabling exceptions where useful, but that is going to be it.  The exception support has its uses, and the way it is currently set up, it makes it possible to use it without getting in the way of "normal" use of LAMMPS. The same kind of request (why is flag X not enabled by default since it is so useful (for me)) could be made about anything, so the bottom lines remains: if you want it, you can enable it.

I think this discussion has gone beyond its usefulness and I will close and lock it now.

**Comment by yurivict** [comment]
Maybe you should compile an executable that should run in parallel without exceptions and without using a shared library, and have a shared library with exceptions for a non-parallel use case only?

**Comment by akohlmey** [comment]
> Maybe you should compile the executable that should run in parallel without exceptions and without using a shared library, and have a shared library with exceptions for a non-parallel use case only?

Why is it such a problem that you need to set an additional flag in the makefile or a CMake variable. You need to do the same for almost any compilation of LAMMPS anyway, so why complain about one more setting? You could just as well complain that not all packages are enabled by default.

**Comment by junghans** [comment]
I had a similar issue in the past! I was wondering what is the performance penalty of enabling the exceptions?

**Comment by yurivict** [comment]
> Why is it such a problem that you need to set an additional flag in the makefile or a CMake variable?

Because I maintain the lammps package on FreeBSD. It doesn't have exceptions enabled by default. But one user, Vipster https://github.com/sgsaenger/vipster, needs exceptions to be enabled. The general package can't be used. This is why I asked here.

The above discussion only convinced me that this is an unsolved problem. "Just rebuild" isn't a choice when you talk about packages. There should either be two packages installable in parallel, or one package should support both use cases.

IMO, the solution might be that the ```lammps``` executable shouldn't link to liblammps.so and shouldn't use exceptions. liblammps.so should be built with exceptions for non-parallel use case.

**Comment by yurivict** [comment]
> I had a similar issue in the past! I was wondering what is the performance penalty of enabling the exceptions?

No performance penalty in this case. This option doesn't disable exceptions, just makes code to not throw them.

**Comment by akohlmey** [comment]
> > Why is it such a problem that you need to set an additional flag in the makefile or a CMake variable?
> 
> Because I maintain the lammps package on FreeBSD. It doesn't have exceptions enabled by default. But one user, Vipster https://github.com/sgsaenger/vipster, needs exceptions to be enabled. The general package can't be used. This is why I asked here.

Sorry, but that is not our problem. You can choose as the package maintainer what to include or enable and what not and how you package it. We welcome people to do that, but we do not develop LAMMPS for package maintainers and whatever requirements, policies, or conventions they have in their distributions or their users ask of them. We try to accommodate what can be done easily, but not at the expense of our primary audience, or when the solution will require significant time and effort on our side. To reiterate, if there is a problem here, it is a packaging problem, not a LAMMPS problem.

**Comment by yurivict** [comment]
The project should support various use cases. As it is now, it can't be packaged for two different use cases simultaneously.

Please let me know how can I compile it such that both parallel users who don't need exceptions and non-parallel users who do need exceptions can be satisfied at the same time?

**Comment by junghans** [comment]
> > I had a similar issue in the past! I was wondering what is the performance penalty of enabling the exceptions?
> 
> No performance penalty in this case. This option doesn't disable exceptions, just makes code to not throw them.

Maybe there are other downsides to enable them, but if there are no performance penalty couldn't we always enable them and then catch all of them in `main.cpp` for the executable use case? That would also get rid of the problem that API changes whether or not the consumer add `-DLAMMPS_EXCEPTIONS` to their code. (Not a problem anymore when using CMake's exported targets)

> IMO, the solution might be that the `lammps` executable shouldn't link to liblammps.so and shouldn't use exceptions. liblammps.so should be built with exceptions for non-parallel use case.

That sounds like a multi-build to me, just build lammps with `BUILD_SHARED_LIBS=OFF` for the executable and `BUILD_SHARED_LIBS=ON` for the library.

**Comment by akohlmey** [comment]
> The project should support various use cases. As it is now, it can't be packaged for two different use cases simultaneously.
> 
> Please let me know how can I compile it such that both parallel users who don't need exceptions and non-parallel users who do need exceptions can be satisfied at the same time?

There are multiple ways to handle conflicting packages that provide the equivalent functionalities. You can use environment modules or Lmod (Fedora/RHEL/CentOS uses that for different MPI libraries and the applications compiled/linked against them), you can use the "alternatives" software that maintains symbolic links, you can use, e.g., `-DLAMMPS_MACHINE=<something>` to have a different library and executable names or you can construct your own way of installing LAMMPS by copying what you want, where you want. I repeat, that is all up to you and not the job of the LAMMPS developers.

**Comment by akohlmey** [comment]
> > > I had a similar issue in the past! I was wondering what is the performance penalty of enabling the exceptions?
> > 
> > 
> > No performance penalty in this case. This option doesn't disable exceptions, just makes code to not throw them.
> 
> Maybe there are other downsides to enable them, but if there are no performance penalty couldn't we always enable them and then catch all of them in `main.cpp` for the executable use case? That would also get rid of the problem that API changes whether or not the consumer add `-DLAMMPS_EXCEPTIONS` to their code. (Not a problem anymore when using CMake's exported targets)

Not for the foreseeable future. Please remember, that CMake is not the only option to build LAMMPS and we want to keep things consistent across the build procedures. And as I've explained earlier in this thread, problems of packaging or people building applications on top of those packages are not something that I have the time or desire to worry about. Many of the core LAMMPS developers don't even use the CMake build process and they don't care.

**Comment by junghans** [comment]
#2090 is what is was thinking of.
